git/pkt_line_writer.go

```diff
@@ -0,0 +1,134 @@
+package git
+
+import (
+	"io"
+
+	"github.com/github/git-lfs/tools"
+)
+
+// PktlineWriter is an implementation of `io.Writer` which writes data buffers
+// "p" to an underlying pkt-line stream for use with the Git pkt-line format.
+type PktlineWriter struct {
+	// buf is an internal buffer used to store data until enough has been
+	// collected to write a full packet, or the buffer was instructed to
+	// flush.
+	buf []byte
+	// pl is the place where packets get written.
+	pl *pktline
+}
+
+var _ io.Writer = new(PktlineWriter)
+
+// NewPktlineWriter returns a new *PktlineWriter, which will write to the
+// underlying data stream "w". The internal buffer is initialized with the given
+// capacity, "c".
+//
+// If "w" is already a `*PktlineWriter`, it will be returned as-is.
+func NewPktlineWriter(w io.Writer, c int) *PktlineWriter {
+	if pw, ok := w.(*PktlineWriter); ok {
+		return pw
+	}
+
+	return &PktlineWriter{
+		buf: make([]byte, 0, c),
+		pl:  newPktline(nil, w),
+	}
+}
+
+// Write implements the io.Writer interface's `Write` method by providing a
+// packet-based backend to the given buffer "p".
+//
+// As many bytes are removed from "p" as possible and stored in an internal
+// buffer until the amount of data in the internal buffer is enough to write a
+// single packet. Once the internal buffer is full, a packet is written to the
+// underlying stream of data, and the process repeats.
+//
+// When the caller has no more data to write in the given chunk of packets, a
+// subsequent call to `Write(p []byte)` MUST be made with a nil slice, to flush
+// the remaining data in the buffer, and write the terminating bytes to the
+// underlying packet stream. Alternatively, the caller can use the public
+// `Flush() error` func to preform the same operation.
```

Comment:
This is becoming a problem. How long until you can stop blaming your new keyboard :trollface: